pjmedia/include/pjmedia-codec.h

```diff
@@ -42,6 +42,8 @@
 #include <pjmedia-codec/speex.h>
 #include <pjmedia-codec/vid_toolbox.h>
 #include <pjmedia-codec/vpx.h>
+#include <pjmedia-codec/and_aud_mediacodec.h>
+#include <pjmedia-codec/and_vid_mediacodec.h>
```

Comment:
> > > > To add AMediaCodec support, include this on the `config_site.h` file.
> > > > ```
> > > > #define PJMEDIA_HAS_ANDROID_MEDIACODEC 1
> > > > ```
> > > 
> > > 
> > > Why not simply have this setting auto-detect & auto-set ?
> > 
> > 
> > Cons of enabling Android MediaCodec by default:
> > 
> > * Most API used are supported only on API 21. Older device will have building issue.
> 
> Not enabled-by-default, but auto-detect (so it will be enabled only when supported, and the auto detect can be turn off via configure script param e.g: --no-android-mediacodec).

Okay. Now it's auto-detect. With new option "--disable-android-mediacodec"

> 
> > * Some codec are supported on newer devices. e.g: VP9 is not yet supported on Android 7.0
> 
> Perhaps the common scenario among app developers is this: use native codecs when supported, otherwise fallback to third-party (e.g: OpenH264, vpx.c). So for each codec, ideally, unless disabled at compile time, the native codec availability is detected on run-time, at compile-time user can still enable both native & third-party (e.g: OpenH264), so when the native impl is not available on the device, PJMEDIA codec will fallback to the third-party impl.
> 
> Perhaps this can be simply done by reordering the codec registration (native codecs are registered first)?

Done.

> 
> > > Usually native codecs are preferred, so if this is stable enough and no compatibility issue with OpenH264, perhaps better to make them (or at least H264) enabled by default?
> > 
> > 
> > Anroid MediaCodec supports software and hardware codec access. Based on tests, the most stable configuration is to use
> > software encoder and hardware decoder. e.g: qcom SoC doesn't provides Exynos or other codec.
> > Currently, the encoder/decoder names are hardcoded and will be enabled if its able to be created.
> > Last test is for qcom (qualcomm) and Exynos SoC.
> > On runtime, the logcat showed this message on encoding process (repeatedly):
> > ```
> > E/NdkMediaCodec: sf error code: -13
> > ```
> > 
> > 
> > However, there's no error was returned from the code and the video was successfully rendered on the remote side.
> 
> Please clarify about 'software encoder' and 'hardware decoder', how such configuration (or any other configuration such as all software/hardware) can be configured?

The codec name indicates it is a software/hardware encoder/decoder.
e.g: codec prefix with "OMX.google" or "c2.android" is software codec.
Hardware decoder contains the SoC names on the prefix.
e.g: "OMX.qcom." or "OMX.exynos." 

https://stackoverflow.com/questions/37715529/how-to-know-android-decoder-mediacodec-createdecoderbytypetype-is-hardware-or

Reply:
So is it possible that a device supports all combinations, i.e: hardware & software for both encoder & decoder? If yes, how an app that wants to have 'the most stable configuration', which mentioned above: software encoder and hardware decoder, can configure that?